### PR TITLE
[docs] Revamp FairScale documentation

### PR DESCRIPTION
**Pull request by anj-s**
## What does this PR do?
- Adds deep dives and tutorials for all APIs (barring a few that are experimental).
- Modify the layout and general structure of the site.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Review by min-xu-ai — APPROVED** [review]
LGTM

Thanks for this great effort!

File: docs/source/deep_dive/activation_checkpointing.rst [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+large activation tensors which allows us to increase the batch size and thereby the net throughput 
+of the model.
+
+
+Activation checkpointing is implemented by overriding `torch.autograd.Function`. In the `forward` 
+function which handles the forward pass of the module, using `no_grad`, we can prevent the creation 
+of the forward graph and materialization of intermediate activation tensors for a long period of 
+time (i.e till the backward pass). Instead, during the backward pass, the forward pass is executed 
+again followed by the backward pass. The inputs to the forward pass are saved using a context object 
+that is then accessed in the backward pass to retrieve the original inputs. We also save the RNG
```
Comment (Vittorio-Caggiano) [inline docs/source/deep_dive/activation_checkpointing.rst:17]: can we spell out RNG?

**Review by Vittorio-Caggiano — COMMENTED** [review]
overall lgtm!

**Comment by SeanNaren** [comment]
This is really nice @anj-s and clears a lot of stuff up with great explanations, thanks for this :)

**Pull request merged**